corehq/apps/commtrack/parsing.py

```diff
@@ -0,0 +1,248 @@
+
+- Individual Ledger Balance:
+
+    <balance xmlns="http://commcarehq.org/ledger/v1" entity-id="" date="" section-id="">
+        <entry id="" quantity="" /> <!--multiple-->
+    </balance>
+
+  parsed into JSON as
+
+    {"@xmlns": "http://commcarehq.org/ledger/v1", "@entity-id": "",
+     "@date": "", "@section-id": "",
+     "entry": [{"@id": "", "@quantity": ""}]}
+
+- Per-Entry Ledger Balance:
+
+    <balance xmlns="http://commcarehq.org/ledger/v1" entity-id="" date="">
+        <entry id=""> <!--multiple-->
+            <value section-id="" quantity=""/> <!-- multiple -->
+        </entry>
+    </balance>
+
+  parsed into JSON as
+
+    {"@xmlns": "http://commcarehq.org/ledger/v1", "@entity-id": "",
+     "@date": "",
+     "entry": [{"@id": "", "value": [{"@section-id: "", "@quantity": ""},
+               ...]}]}
+
+Conceptually, both formats produce a list of transactions:
+
+    (entity_id="", date="", section_id="", entry_id="", quantity="")
+    ...
+
+but Per-Entry lets you have many different section_ids among the transactions.
+
+"""
+from collections import namedtuple
+from decimal import Decimal
+import logging
+import datetime
+from casexml.apps.stock import const as stockconst
+from casexml.apps.stock.const import COMMTRACK_REPORT_XMLNS
+from corehq.apps.commtrack import const
+from corehq.apps.commtrack.exceptions import InvalidDate
+from corehq.apps.commtrack.models import StockReportHelper, \
+    StockTransactionHelper
+from couchforms.models import XFormInstance
+from couchforms.util import adjust_datetimes
+from xml2json.lib import convert_xml_to_json
+
+
+LedgerInstruction = namedtuple(
+    'LedgerInstruction',
+    ['date', 'section_id', 'entry_id', 'quantity',
+     # for balance instructions
+     'entity_id',
+     # for transfer instructions
+     'src', 'dest', 'type']
+)
+
+
+class LedgerFormat(object):
+    individual = object()
+    per_entry = object()
+
+
+def should_be_a_list(obj_or_list):
+    if obj_or_list is None:
+        return []
+    elif isinstance(obj_or_list, list):
+        return obj_or_list
+    else:
+        return [obj_or_list]
+
+
+def unpack_commtrack(xform):
+    form_xml = xform.get_xml_element()
+    commtrack_node_names = ('{%s}balance' % COMMTRACK_REPORT_XMLNS,
+                            '{%s}transfer' % COMMTRACK_REPORT_XMLNS)
+
+    def commtrack_nodes(node):
+        for child in node:
+            if child.tag in commtrack_node_names:
+                yield child
+            else:
+                for e in commtrack_nodes(child):
+                    yield e
+
+    for elem in commtrack_nodes(form_xml):
+        report_type, ledger_json = convert_xml_to_json(
+            elem, last_xmlns=COMMTRACK_REPORT_XMLNS)
+
+        # apply the same datetime & string conversions
+        # that would be applied to XFormInstance.form
+        adjust_datetimes(ledger_json)
+        ledger_json = XFormInstance({'form': ledger_json}).form
```

Comment:
Yeah, it initially caught the attention of my "too clever" sensor, but I decided what it's saying exactly semantically (and not just functionally) correct: "Please wrap this field exactly as `XFormInstance` would wrap its `form` field".